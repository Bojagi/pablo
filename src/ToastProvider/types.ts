import { ToastCardProps } from '../ToastCard';

export interface InternalToastMessageOptions extends ToastMessageOptions {
  hidden: boolean;
}

export type ToastStackSide = 'bottom-right' | 'bottom-left' | 'top-right' | 'top-left';

export interface ToastMessageOptions extends ToastCardProps {
  id?: string;
  duration: number;
}
