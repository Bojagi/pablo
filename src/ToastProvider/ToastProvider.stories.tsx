import React, { useCallback, useState } from 'react';
import { Box, Flex } from '../Box';
import { Button } from '../Button';
import { Checkbox } from '../Checkbox';
import { Input } from '../Input';
import { ToastProvider, useToast } from './ToastProvider';

export default {
  title: 'ToastProvider',
};

const Trigger = () => {
  const { addToast } = useToast();
  const [counter, setCounter] = useState(1);
  const [closable, setClosable] = useState(true);
  const [duration, setDuration] = useState(3000);

  const handleAddToast = useCallback(() => {
    addToast({
      duration,
      title: `Hello there #${counter}`,
      description: 'Something happened!',
      type: 'success',
      closable,
    });
    setCounter(counter + 1);
  }, [addToast, duration, closable, counter]);

  return (
    <Flex width="100%" height="90vh" justifyContent="center" alignItems="center">
      <Box>
        <Checkbox
          mb={4}
          label="closable"
          checked={closable}
          onChange={() => setClosable((val) => !val)}
        />
        <Input
          label="Duration"
          mb={4}
          onChange={(v) => setDuration(parseInt(v, 10) || 0)}
          value={duration}
        />
        <Button onClick={handleAddToast}>Add toast</Button>
      </Box>
    </Flex>
  );
};

const BaseStory = (args) => (
  <ToastProvider {...args}>
    <Trigger />
  </ToastProvider>
);

export const BottomRight = BaseStory.bind(null);
BottomRight.args = {
  side: 'bottom-right',
};

export const BottomLeft = BaseStory.bind(null);
BottomLeft.args = {
  side: 'bottom-left',
};

export const TopRight = BaseStory.bind(null);
TopRight.args = {
  side: 'top-right',
};

export const TopLeft = BaseStory.bind(null);
TopLeft.args = {
  side: 'top-left',
};
