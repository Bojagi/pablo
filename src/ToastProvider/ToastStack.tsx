import React, { ReactNode } from 'react';
import styled, { css } from 'styled-components';
import { getComponentStyle, getSpacing } from '../utils/styleHelpers';
import { ToastStackSide } from './types';

export interface ToastStackProps {
  side: ToastStackSide;
  children: ReactNode;
}

const ToastStackBox = styled.div<ToastStackProps>`
  padding: ${getSpacing(1.5)};
  width: ${getComponentStyle('toastCard.width')};
  position: fixed;
  display: flex;
  z-index: 1100;
  ${(props) => {
    switch (props.side) {
      case 'bottom-right':
        return css`
          bottom: 0;
          flex-direction: column-reverse;
          right: 0;
        `;
      case 'bottom-left':
        return css`
          bottom: 0;
          flex-direction: column-reverse;
          left: 0;
        `;
      case 'top-right':
        return css`
          top: 0;
          flex-direction: column;
          right: 0;
        `;
      case 'top-left':
        return css`
          top: 0;
          flex-direction: column;
          left: 0;
        `;
      default:
        return null;
    }
  }}
`;

export function ToastStack({ children, side }: ToastStackProps) {
  return (
    <ToastStackBox data-testid="pbl-toaststack" side={side}>
      {children}
    </ToastStackBox>
  );
}
