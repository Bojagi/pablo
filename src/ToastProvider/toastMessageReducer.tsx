import { InternalToastMessageOptions, ToastMessageOptions } from './types';

interface AddToastAction {
  type: 'add';
  message: ToastMessageOptions;
}

interface HideToastAction {
  type: 'hide';
  id: string;
}

interface RemoveToastAction {
  type: 'remove';
  id: string;
}

export function toastMessageReducer(
  state: InternalToastMessageOptions[],
  action: AddToastAction | RemoveToastAction | HideToastAction
) {
  switch (action.type) {
    case 'add':
      return addMessage(state, action.message);
    case 'hide':
      return hideMessage(state, action.id);
    case 'remove':
      return removeMessage(state, action.id);
    default:
      throw new Error();
  }
}

function addMessage(state: InternalToastMessageOptions[], message: ToastMessageOptions) {
  return [
    ...state,
    {
      ...message,
      hidden: false,
    },
  ];
}

function hideMessage(state: InternalToastMessageOptions[], id: string) {
  const indexToHide = state.findIndex((item) => item.id === id);
  const messages = [...state];
  const updatedMessage: InternalToastMessageOptions = {
    ...messages[indexToHide],
    hidden: true,
  };
  messages.splice(indexToHide, 1, updatedMessage);
  return messages;
}

function removeMessage(state: InternalToastMessageOptions[], id: string) {
  const indexToDelete = state.findIndex((item) => item.id === id);
  const messages = [...state];
  messages.splice(indexToDelete, 1);
  return messages;
}
