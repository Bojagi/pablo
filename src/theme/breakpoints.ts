export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
