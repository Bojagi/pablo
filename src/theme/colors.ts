export type ShadedColor = {
  main: string;
  dark: string;
  darkest: string;
  light: string;
  lightest: string;
  contrastText: string;
  contrastTextLight: string;
};

export type CommonColors = {
  white: string;
  black: string;
  whiteContrastText: string;
  blackContrastText: string;
};

export type TextColors = {
  main: string;
  info: string;
};

export type BorderColors = {
  main: string;
  lightest: string;
};

export interface GrayColors {
  '50': string;
  '100': string;
  '200': string;
  '300': string;
  '400': string;
  '500': string;
  '600': string;
  '700': string;
  '800': string;
  '900': string;
}

export type AllColors = ShadedColor & TextColors & BorderColors & CommonColors;

export interface Colors {
  common: CommonColors;
  gray: GrayColors;
  borders: BorderColors;
  brand: ShadedColor;
  positive: ShadedColor;
  negative: ShadedColor;
  neutral: ShadedColor;
  text: TextColors;
}

export const WHITE = '#ffffff';
export const BLACK = '#000000';

export const colors: Colors = {
  common: {
    white: WHITE,
    black: BLACK,
    whiteContrastText: BLACK,
    blackContrastText: WHITE,
  },
  gray: {
    '50': getGrayColor(0.95),
    '100': getGrayColor(0.9),
    '200': getGrayColor(0.8),
    '300': getGrayColor(0.7),
    '400': getGrayColor(0.6),
    '500': getGrayColor(0.5),
    '600': getGrayColor(0.4),
    '700': getGrayColor(0.3),
    '800': getGrayColor(0.2),
    '900': getGrayColor(0.1),
  },
  borders: {
    main: `${BLACK}${getHexByte(0.25)}`,
    lightest: `${BLACK}${getHexByte(0.1)}`,
  },
  text: {
    main: BLACK,
    info: `${BLACK}${getHexByte(0.5)}`,
  },
  brand: {
    lightest: '#EBE7FF',
    light: '#D6CEFF',
    main: '#6A50F2',
    dark: '#4129BD',
    darkest: '#281683',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  positive: {
    lightest: '#DFF6E6',
    light: '#BBECCB',
    main: '#38C765',
    dark: '#2FB85A',
    darkest: '#108B36',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  negative: {
    lightest: '#FFE7E5',
    light: '#FFCFCC',
    main: '#F21D0D',
    dark: '#CA271B',
    darkest: '#871A12',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  neutral: {
    lightest: '#FCF5DB',
    light: '#FFEB9E',
    main: '#FFDD5B',
    dark: '#EDC52A',
    darkest: '#D6AC0B',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
};

/**
 * Get hex color byte (e.g. to get transparency)
 * @param value Percentage value (between 0.0 and 1.0)
 */
function getHexByte(value: number) {
  return (Math.round((255 * value) / 16) * 16).toString(16);
}

function getGrayColor(value: number) {
  return `#${getHexByte(value)}${getHexByte(value)}${getHexByte(value)}`;
}
