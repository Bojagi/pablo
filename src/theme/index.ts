export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
