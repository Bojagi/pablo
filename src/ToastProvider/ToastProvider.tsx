import React, { ReactNode, useReducer } from 'react';
import { Box } from '../Box';
import { ToastCard } from '../ToastCard';
import { createUniqueId } from '../utils/useUniqueId';
import { StackAnimation } from '../animation/StackAnimation';
import { toastMessageReducer } from './toastMessageReducer';
import { ToastStack } from './ToastStack';
import { ToastMessageOptions, ToastStackSide } from './types';

const DEFAULT_DURATION = 3000;
export const FADE_DURATION = 300;

export type AddMessageFunction = (messageOptions: ToastMessageOptions) => void;

const addMessageContext = React.createContext<AddMessageFunction>(() => {});

export function useToast() {
  const addToast = React.useContext(addMessageContext);
  return { addToast };
}

export interface ToastProviderProps {
  side?: ToastStackSide;
  children: ReactNode;
}

export function ToastProvider({ children, side = 'bottom-right' }: ToastProviderProps) {
  const [messages, dispatchMessage] = useReducer(toastMessageReducer, []);

  const addMessage = React.useCallback(
    (message: ToastMessageOptions) =>
      dispatchMessage({
        type: 'add',
        message,
      }),
    []
  );

  const removeMessage = React.useCallback((id: string) => {
    dispatchMessage({
      type: 'remove',
      id,
    });
  }, []);

  const hideMessage = React.useCallback((id: string) => {
    dispatchMessage({
      type: 'hide',
      id,
    });
  }, []);

  const handleAddMessage = React.useCallback(
    (messageOptions: ToastMessageOptions) => {
      const id = messageOptions.id || createUniqueId('toast-message');
      const duration = messageOptions.duration ?? DEFAULT_DURATION;
      if (duration) {
        setTimeout(() => hideMessage(id), duration);
      }
      addMessage({
        ...messageOptions,
        id,
      });
    },
    [hideMessage, addMessage]
  );

  return (
    <>
      <addMessageContext.Provider value={handleAddMessage}>{children}</addMessageContext.Provider>
      <ToastStack side={side}>
        {messages.map(({ hidden, ...message }, i, arr) => (
          <Box key={message.id} zIndex={arr.length - i}>
            <StackAnimation
              duration={FADE_DURATION}
              visible={!hidden}
              onExited={() => removeMessage(message.id!)}
            >
              <ToastCard {...message} mb={1} onClose={() => hideMessage(message.id!)} />
            </StackAnimation>
          </Box>
        ))}
      </ToastStack>
    </>
  );
}
