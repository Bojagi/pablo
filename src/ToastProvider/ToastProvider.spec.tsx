import { act, cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { ToastProvider, useToast } from './ToastProvider';
import '../../testUtils/mockResizeObserver';

beforeEach(() => {
  jest.useFakeTimers();
});

afterEach(cleanup);

test('Render toast stack on the bottom-right by default', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
  });
  expect(getByTestId('pbl-toaststack')).toMatchSnapshot();
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('left', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('bottom', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('right', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('flex-direction', 'column-reverse');
});

test('Render toast stack on the bottom-right', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'bottom-right',
  });
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('bottom', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('right', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('flex-direction', 'column-reverse');
});

test('Render toast stack on the bottom-left', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'bottom-left',
  });
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('left', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('bottom', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('right', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('flex-direction', 'column-reverse');
});

test('Render toast stack on the top-left', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-left',
  });
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('left', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('bottom', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('right', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('flex-direction', 'column');
});

test('Render toast stack on the top-right', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-right',
  });
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('top', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('left', '0');
  expect(getByTestId('pbl-toaststack')).not.toHaveStyleRule('bottom', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('right', '0');
  expect(getByTestId('pbl-toaststack')).toHaveStyleRule('flex-direction', 'column');
});

test('Add a message to toast stack', () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
  });

  expect(getByTestId('pbl-toastcard-title')).toHaveTextContent('Hello there');
  expect(getByTestId('pbl-toastcard-description')).toHaveTextContent('Something happened!');
});

test('Add multiple message to toast stack', () => {
  const { getByTestId, getAllByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
    fireEvent.click(getByTestId('add-toast'));
  });

  expect(getAllByTestId('pbl-toastcard-title')[0]).toHaveTextContent('Hello there');
  expect(getAllByTestId('pbl-toastcard-description')[0]).toHaveTextContent('Something happened!');
  expect(getAllByTestId('pbl-toastcard-title')[1]).toHaveTextContent('Hello there');
  expect(getAllByTestId('pbl-toastcard-description')[1]).toHaveTextContent('Something happened!');
});

test.skip('Hide message after duration and animation', async () => {
  const { getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
    // introduction animation (300ms) plus 100ms visibility
    jest.advanceTimersByTime(400);
  });

  // Trigger leaving animation by waiting for RAF
  await act(() => new Promise((resolve) => requestAnimationFrame(resolve as any)));
  act(() => {
    jest.advanceTimersByTime(299);
  });

  // Should still be there as the animation takes 300ms
  expect(getByTestId('pbl-toastcard-title')).toHaveTextContent('Hello there');
  expect(getByTestId('pbl-toastcard-description')).toHaveTextContent('Something happened!');

  act(() => {
    jest.advanceTimersByTime(1);
  });

  await waitFor(
    () => {
      expect(getByTestId('pbl-toaststack').childNodes).toHaveLength(0);
    },
    {
      timeout: 500,
    }
  );
});

test.skip('Hide message after default duration and animation', async () => {
  const { getByTestId } = renderComponent({
    type: 'success',
    closable: true,
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
    jest.advanceTimersByTime(3300);
  });

  // Trigger leaving animation by waiting for RAF
  await act(() => new Promise((resolve) => requestAnimationFrame(resolve as any)));
  act(() => {
    jest.advanceTimersByTime(300);
  });

  expect(getByTestId('pbl-toaststack').childNodes).toHaveLength(0);
});

test('Hide message after duration and animation with component being unmounted in the meantime', () => {
  const { unmount, getByTestId } = renderComponent({
    duration: 100,
    type: 'success',
    closable: true,
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
    jest.advanceTimersByTime(100);
  });

  unmount();

  act(() => {
    jest.advanceTimersByTime(300);
  });
});

test.skip('Should not hide message after time when duration is 0', async () => {
  const { getByTestId } = renderComponent({
    duration: 0,
    type: 'success',
    side: 'top-right',
  });

  act(() => {
    fireEvent.click(getByTestId('add-toast'));
    jest.advanceTimersByTime(1000);
  });

  // Should still be there as duration is 0
  expect(getByTestId('pbl-toastcard-title')).toHaveTextContent('Hello there');
  expect(getByTestId('pbl-toastcard-description')).toHaveTextContent('Something happened!');

  // Trigger leaving animation by clicking and waiting for RAF
  act(() => {
    fireEvent.click(getByTestId('pbl-toastcard-closebtn'));
  });
  await act(() => new Promise((resolve) => requestAnimationFrame(resolve as any)));
  act(() => {
    jest.advanceTimersByTime(299);
  });

  // Should still be there as the animation takes 300ms
  expect(getByTestId('pbl-toastcard-title')).toHaveTextContent('Hello there');
  expect(getByTestId('pbl-toastcard-description')).toHaveTextContent('Something happened!');

  act(() => {
    jest.advanceTimersByTime(1);
  });

  expect(getByTestId('pbl-toaststack').childNodes).toHaveLength(0);
});

function Trigger({ duration, type, closable }) {
  const { addToast } = useToast();

  const handleAddToast = React.useCallback(() => {
    addToast({
      duration,
      title: `Hello there`,
      description: 'Something happened!',
      type,
      closable,
    });
  }, [addToast, closable, type, duration]);

  return <button data-testid="add-toast" onClick={handleAddToast}></button>;
}

function renderComponent({ duration, type, closable, ...props }: any) {
  return render(
    <PabloThemeProvider>
      <ToastProvider {...props}>
        <Trigger duration={duration} type={type} closable={closable} />
      </ToastProvider>
    </PabloThemeProvider>
  );
}
