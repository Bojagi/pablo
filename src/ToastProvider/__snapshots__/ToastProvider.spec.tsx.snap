// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render toast stack on the bottom-right by default 1`] = `
.emotion-0 {
  box-sizing: border-box;
  padding: clamp(0.75rem, 0.6428571428571428rem + 0.5357142857142857vw, 1.125rem);
  width: 400px;
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 1100;
  bottom: 0;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
  right: 0;
}

<div
  class="emotion-0"
  data-testid="pbl-toaststack"
/>
`;
